src/System.Linq/src/System/Linq/Reverse.cs

```diff
@@ -12,13 +14,88 @@ public static partial class Enumerable
+            return new ReverseIterator<TSource>(source);
         }
 
-        private static IEnumerable<TSource> ReverseIterator<TSource>(IEnumerable<TSource> source)
+        private sealed class ReverseIterator<TSource> : Iterator<TSource>, IIListProvider<TSource>
         {
-            Buffer<TSource> buffer = new Buffer<TSource>(source);
-            for (int i = buffer.count - 1; i >= 0; i--) yield return buffer.items[i];
+            private readonly IEnumerable<TSource> _source;
+            private TSource[] _buffer;
+            private int _index;
+
+            public ReverseIterator(IEnumerable<TSource> source)
+            {
+                Debug.Assert(source != null);
+                _source = source;
+            }
+
+            public override Iterator<TSource> Clone()
+            {
+                return new ReverseIterator<TSource>(_source);
+            }
+
+            public override bool MoveNext()
+            {
+                switch (state)
+                {
+                    case 1:
+                        Buffer<TSource> buffer = new Buffer<TSource>(_source);
+                        _buffer = buffer.items;
+                        _index = buffer.count - 1;
+                        state = 2;
+                        goto case 2;
+                    case 2:
+                        if (_index != -1)
+                        {
+                            current = _buffer[_index];
+                            --_index;
+                            return true;
+                        }
+                        break;
+                }
+
+                Dispose();
+                return false;
+            }
+
+            public override void Dispose()
+            {
+                _buffer = null; // Just in case this ends up being long-lived, allow the memory to be reclaimed.
+                base.Dispose();
+            }
+
+            public TSource[] ToArray()
+            {
+                TSource[] array = _source.ToArray();
+                // Array.Reverse() involves boxing for non-primitive value types, but
+                // checking that has its own cost, so just use this approach for all types.
+                for (int i = 0, j = array.Length - 1; i < j; ++i, --j)
+                {
+                    TSource temp = array[i];
+                    array[i] = array[j];
+                    array[j] = temp;
+                }
+                return array;
+            }
+
+            public List<TSource> ToList()
+            {
+                List<TSource> list = _source.ToList();
+                list.Reverse();
+                return list;
+            }
+
+            public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                    if (listProv != null) return listProv.GetCount(onlyIfCheap: true);
```

Comment:
@stephentoub likewise this line differs to what you'd already reviewed in the same way.